File: C++/Telemaster/aero.cpp
#include "aero.h"

aero::aero()
{
    memset(&y, 0, sizeof(y));
    memset(&u, 0, sizeof(u));
    memset(&internal, 0, sizeof(internal));
}

void aero::update() 
{   
    // Wind
    internal.V_rel = *u.v_b - *u.Wind;
    
    // Aerodynamics System
    internal.V = u.v_b->norm();
    internal.alpha = atan2(internal.V_rel.z(), internal.V_rel.x());
    internal.beta = asin(internal.V_rel.y() /internal.V);
    internal.Cba = AngleAxisd(internal.alpha, Vector3d(0.0,-1.0,0.0)) * AngleAxisd(internal.beta, Vector3d(0.0, 0.0, -1.0));

    // Aerodynamics Derivatives
    internal.CL0  = 0.3647;
    internal.CM0  = -0.0031;
    internal.CD0  = 0.05917;
    internal.CDde =  2E-07 * pow(internal.alpha,2)  - 2E-05 * internal.alpha - 0.0007;
    internal.CLa  = -0.0017 * pow(internal.alpha, 2) - 0.0126 * internal.alpha + 4.4523;
    internal.CLde = -4E-06 * pow(internal.alpha, 2) - 2E-06 * internal.alpha + 0.0101;
    internal.CLq  = -0.001 * pow(internal.alpha, 2) - 0.004 * internal.alpha + 8.0732;
    internal.CMa  = 0.0005 * pow(internal.alpha, 2) + 9E-05 * internal.alpha - 0.9113;
    internal.CMde = 7E-06 * pow(internal.alpha, 2) - 2E-07 * internal.alpha - 0.0221;
    internal.CMq  = 0.0013 * pow(internal.alpha, 2) - 1E-05 * internal.alpha - 8.782;
    internal.CDda = 1E-05 * internal.alpha;
    internal.CDdr = 5E-05 * internal.alpha;
    internal.CYb  = 0.0002 * pow(internal.alpha, 2) - 0.3462;
    internal.CYda = 0.000289;
    internal.CYdr = -0.003914;
    internal.CYp  = -1E-05 * pow(internal.beta, 2) + 0.0943;
    internal.CYr  = -5E-05 * pow(internal.beta, 2) + 0.2988;
    internal.Clb  = 3E-05 * pow(internal.beta, 2) - 0.0488;
    internal.Clda = -0.003809;
    internal.Cldr = -0.000165;
    internal.Clp  = 6E-05 * pow(internal.beta, 2) - 0.3977;
    internal.Clr  = -2E-05 * pow(internal.beta, 2) + 0.1078;
    internal.Cnb  = -8E-05 * pow(internal.beta, 2) + 0.1308;
    internal.Cnda = 0.00011;
    internal.Cndr = 0.001833;
    internal.Cnp  = 4E-06 * pow(internal.beta, 2) - 0.0291;
    internal.Cnr  = 2E-05 * pow(internal.beta, 2) - 0.1366;
    
    // Aircraft Geometry
    internal.S = 0.9324;
    internal.c = 0.42;
    internal.b = 2.220;
    internal.Aw = 5.2857;
    internal.k = 0.0492;
    
    if (internal.V < 10.0)
    {
        internal.CL = 0.0;
        internal.CD = 0.0;
        internal.CM = 0.0;
        internal.CY = 0.0;
        internal.Cl = 0.0;
        internal.Cn = 0.0;
    }
    else
    {
        internal.CL = internal.CL0 + internal.CLa * internal.alpha + internal.CLde * ((*u.cmd)(1) * 180.0 / M_PI) + internal.CLq * ((u.w_b->y() * internal.c) / (2.0 * internal.V));
        internal.CD = internal.CD0 + internal.CDda * ((*u.cmd)(0) * 180.0 / M_PI) + internal.CDde * ((*u.cmd)(1) * 180.0 / M_PI) + internal.CDdr * ((*u.cmd)(2) * 180.0 / M_PI) + pow(internal.CL, 2) * internal.k;
        internal.CM = internal.CM0 + internal.CMa * internal.alpha + internal.CMde * ((*u.cmd)(1) * 180.0 / M_PI) + internal.CMq * ((u.w_b->y() * internal.c) / (2.0 * internal.V));
        internal.CY = internal.CYb * internal.beta + internal.CYda * ((*u.cmd)(0) * 180.0 / M_PI) + internal.CYdr * ((*u.cmd)(2) * 180.0 / M_PI) + internal.CYp * ((u.w_b->x() * internal.b) / (2.0 * internal.V)) + internal.CYr * ((u.w_b->z() * internal.b) / (2.0 * internal.V));
        internal.Cl = internal.Clb * internal.beta + internal.Clda * ((*u.cmd)(0) * 180.0 / M_PI) + internal.Cldr * ((*u.cmd)(2) * 180.0 / M_PI) + internal.Clp * ((u.w_b->x() * internal.b) / (2.0 * internal.V)) + internal.Clr * ((u.w_b->z() * internal.b) / (2.0 * internal.V));
        internal.Cn = internal.Cnb * internal.beta + internal.Cnda * ((*u.cmd)(0) * 180.0 / M_PI) + internal.Cndr * ((*u.cmd)(2) * 180.0 / M_PI) + internal.Cnp * ((u.w_b->x() * internal.b) / (2.0 * internal.V)) + internal.Cnr * ((u.w_b->z() * internal.b) / (2.0 * internal.V));
    }

    internal.D = 0.5 * (*u.rho) * pow(internal.V, 2) * internal.CD * internal.S;
    internal.Y = 0.5 * (*u.rho) * pow(internal.V, 2) * internal.CY * internal.S;
    internal.L = 0.5 * (*u.rho) * pow(internal.V, 2) * internal.CL * internal.S;
    internal.l = 0.5 * (*u.rho) * pow(internal.V, 2) * internal.Cl * internal.S * internal.b;
    internal.M = 0.5 * (*u.rho) * pow(internal.V, 2) * internal.CM * internal.S * internal.c;
    internal.N = 0.5 * (*u.rho) * pow(internal.V, 2) * internal.Cn * internal.S * internal.b;
}

void aero::get_output() 
{ 
    y.alpha = internal.alpha;
    y.beta  = internal.beta;
    y.V = internal.V;
    y.CL = internal.CL;
    y.CD = internal.CD;
    y.CM = internal.CM;
    y.Faero_w << internal.D, internal.Y, internal.L;
    y.Faero_b = internal.Cba * -y.Faero_w;
    y.Maero_b << internal.l, internal.M, internal.N;
    y.Cba << internal.Cba;
}

File: MATLAB/AED/aero.m
%% Load
load('AVL/Results/LatCoeffs.mat')
load('AVL/Results/LonCoeffs.mat')
%% Salvar
% Definindo os valores de Alpha e as outras colunas
Alpha = TableAlpha.Alpha;
CL0 = 0.3647;
CD0 = 0.05917;
CM0 = -0.00310;
CLa = TableAlpha.CLa;
Cma = TableAlpha.Cma;
CLq = TableAlpha.CLq;
Cmq = TableAlpha.Cmq;
CLde = TableAlpha.CLd02;
Cmde = TableAlpha.Cmd02;

% Criando uma grade para interpolação
alpha = linspace(min(Alpha), max(Alpha), 100); % Altere o número de pontos conforme necessário

% Interpolando os valores para cada coluna
CLa = interp1(Alpha, CLa, alpha, 'spline');
CMa = interp1(Alpha, Cma, alpha, 'spline');
CLq = interp1(Alpha, CLq, alpha, 'spline');
CMq = interp1(Alpha, Cmq, alpha, 'spline');
CLde = interp1(Alpha, CLde, alpha, 'spline');
CMde = interp1(Alpha, Cmde, alpha, 'spline');

% Definindo os valores de Beta e as outras colunas
Beta = [-10; -5; 0; 5; 10];
Cldr = -1.65e-04;
Clda = -0.003694;
CYdr = -0.003885;
CYda = 2.80e-04;
Cndr = 0.001778;
Cnda = 1.1e-04;
CYb = TableBeta.CYb';
Clb = TableBeta.Clb';
Cnb = TableBeta.Cnb';
CYp = TableBeta.CYp';
Clp = TableBeta.Clp';
Cnp = TableBeta.Cnp';
CYr = TableBeta.CYr';
Clr = TableBeta.Clr';
Cnr = TableBeta.Cnr';

% Criando uma grade para interpolação
beta = linspace(min(Beta), max(Beta), 100); % Altere o número de pontos conforme necessário

% Interpolando os valores para cada coluna
CYb = interp1(Beta, CYb, beta, 'spline');
Clb = interp1(Beta, Clb, beta, 'spline');
Cnb = interp1(Beta, Cnb, beta, 'spline');
CYp = interp1(Beta, CYp, beta, 'spline');
Clp = interp1(Beta, Clp, beta, 'spline');
Cnp = interp1(Beta, Cnp, beta, 'spline');
CYr = interp1(Beta, CYr, beta, 'spline');
Clr = interp1(Beta, Clr, beta, 'spline');
Cnr = interp1(Beta, Cnr, beta, 'spline');


%% Hinge Moments
% elevator
Hmele = [-20 20].';
Hmalpha = [-10 15].';
HM_ele = [-4.37E-03 -5.53E-03
          5.88E-03  4.33E-03];

% rudder
Hmrud = [-20 20].';
Hmbeta = [-10 10].';
HM_rud = [4.14E-03  2.64E-03
         -2.64E-03 -4.14E-03];

% aileron
HM_ail = [-0.00458 0.00458];

save('alpha.mat','alpha','-v7.3')
save('Hmalpha.mat','Hmalpha','-v7.3')
save('beta.mat','beta','-v7.3')
save('Hmbeta.mat','Hmbeta','-v7.3')
save('CL0.mat','CL0','-v7.3') 
save('CM0.mat','CM0','-v7.3') 
save('CD0.mat','CD0','-v7.3') 
save('CLa.mat','CLa','-v7.3') 
save('CLde.mat','CLde','-v7.3') 
save('CLq.mat','CLq','-v7.3') 
save('CMa.mat','CMa','-v7.3') 
save('CMde.mat','CMde','-v7.3') 
save('CMq.mat','CMq','-v7.3')  
save('CYb.mat','CYb','-v7.3') 
save('CYda.mat','CYda','-v7.3') 
save('CYdr.mat','CYdr','-v7.3') 
save('CYp.mat','CYp','-v7.3') 
save('CYr.mat','CYr','-v7.3') 
save('Clb.mat','Clb','-v7.3') 
save('Clda.mat','Clda','-v7.3') 
save('Cldr.mat','Cldr','-v7.3') 
save('Clp.mat','Clp','-v7.3') 
save('Clr.mat','Clr','-v7.3') 
save('Cnb.mat','Cnb','-v7.3') 
save('Cnda.mat','Cnda','-v7.3') 
save('Cndr.mat','Cndr','-v7.3') 
save('Cnp.mat','Cnp','-v7.3') 
save('Cnr.mat','Cnr','-v7.3')
save('Hmele.mat','Hmele','-v7.3')
save('Hmrud.mat','Hmrud','-v7.3')
save('HM_ele.mat','HM_ele','-v7.3')
save('HM_rud.mat','HM_rud','-v7.3')
save('HM_ail.mat','HM_ail','-v7.3')
